File: tutorial_for_Python_and_R/3_multiple_linear_regression.R
# Multiple Linear Regression

# Set the directory
setwd('C:/Users/KYY/Desktop/ML_Ato_Z/[0]Data')

# Importing the dataset
dataset = read.csv('50_Startups.csv')


# Encoding categorical data
dataset$State = factor(dataset$State,
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1, 2, 3))


# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)   # short cut: F1 for help
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)


# Feature Scaling - included in Linear regression package


# Fitting Multiple Linear Regression to the Training set
# 'F1': short cut for help
regressor = lm(formula = Profit ~ .,  # use all the independent variable 
               data = training_set)
summary(regressor)


# Predicting the Test set results
y_pred = predict(regressor, newdata = test_set)



# Building the optimal model using Backward Elimination
# use total dataset (not only trainig set)
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend + State,  # use all the independent variable 
               data = dataset)
summary(regressor)
# => remove all the state (highest p-value over significant level)
#    (because State2: 0.990, State3: 0.943 => state variables seems not to have relationship with the dependent variable)
regressor = lm(formula = Profit ~ R.D.Spend + Administration + Marketing.Spend,  # use all the independent variable 
               data = dataset)
summary(regressor)
# => remove Administration (highest p-value over significant level)
regressor = lm(formula = Profit ~ R.D.Spend + Marketing.Spend,  # use all the independent variable 
               data = dataset)
summary(regressor)
# => remove Marketing.Spend (highest p-value over significant level)
regressor = lm(formula = Profit ~ R.D.Spend,  # use all the independent variable 
               data = dataset)
summary(regressor)


# [Automatic implementation] Backward elimination
backwardElimination <- function(x, sl) {
  numVars = length(x)
  for (i in c(1:numVars)){
    regressor = lm(formula = Profit ~ ., data = x)
    maxVar = max(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"])
    # find that variable
    if (maxVar > sl){
      j = which(coef(summary(regressor))[c(2:numVars), "Pr(>|t|)"] == maxVar)
      x = x[, -j]
    }
    numVars = numVars - 1
  }
  return(summary(regressor))
}

SL = 0.05
dataset = dataset[, c(1,2,3,4,5)]
backwardElimination(training_set, SL)

File: tutorial_for_Python_and_R/3_multiple_linear_regression.py
# -*- coding: utf-8 -*-

# Multiple Linear Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

import os

# Set the directory
currentPath = os.getcwd()
print(currentPath)
os.chdir('C:\\Users\\KYY\\Desktop\\ML_Ato_Z\\[0]Data')


# Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, -1].values


# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])  # 0: index of categorical column
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the dummy Variable Trap
X = X[:, 1:]   # remove 1st column of X


# Splitting the dataset into the Training set and testing set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2,
                                                    random_state = 0)

# Feature Scaling -X- : included in library


# Fitting multiple linear regression to the training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)


# Predicting the Test set results
y_pred = regressor.predict(X_test)


# Building the optimal model using Backward Elimination
import statsmodels.formula.api as sm
X = np.append(arr = np.ones((50, 1)).astype(int), values = X, axis = 1)  # specifc part for statmodel library / constant variable
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()  # Ordinary Least Square
regressor_OLS.summary()
# => remove X2   (highest p-value over significant level 0.05)
X_opt = X[:, [0, 1, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()  # Ordinary Least Square
regressor_OLS.summary()
# => remove X1   (highest p-value over significant level 0.05)
X_opt = X[:, [0, 3, 4, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()  # Ordinary Least Square
regressor_OLS.summary()
# => remove X2   (highest p-value over significant level 0.05)
X_opt = X[:, [0, 3, 5]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()  # Ordinary Least Square
regressor_OLS.summary()
# => remove X2   (highest p-value over significant level 0.05)
X_opt = X[:, [0, 3]]
regressor_OLS = sm.OLS(endog = y, exog = X_opt).fit()  # Ordinary Least Square
regressor_OLS.summary()
# -- Finish


# [Automatic] 1. Backward Elimination with p-values only
# x: data(array), sl: significant level ex) 0.05
import statsmodels.formula.api as sm
def backwardElimination(x, sl):
    numVars = len(x[0])
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues).astype(float)
        if maxVar > sl:
            # find the variable and remove it
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    x = np.delete(x, j, 1)
    regressor_OLS.summary()
    return x
 
SL = 0.05
X_opt = X[:, [0, 1, 2, 3, 4, 5]]  # X: with constant variable column (np.ones())
X_Modeled = backwardElimination(X_opt, SL)


# [Automatic] 2. Backward Elimination with p-values and Adjusted R squared
import statsmodels.formula.api as sm
def backwardElimination(x, SL):
    numVars = len(x[0])
    temp = np.zeros((50,6)).astype(int)
    for i in range(0, numVars):
        regressor_OLS = sm.OLS(y, x).fit()
        maxVar = max(regressor_OLS.pvalues).astype(float)
        adjR_before = regressor_OLS.rsquared_adj.astype(float)
        if maxVar > SL:
            for j in range(0, numVars - i):
                if (regressor_OLS.pvalues[j].astype(float) == maxVar):
                    temp[:,j] = x[:, j]
                    x = np.delete(x, j, 1)
                    tmp_regressor = sm.OLS(y, x).fit()
                    adjR_after = tmp_regressor.rsquared_adj.astype(float)
                    if (adjR_before >= adjR_after):
                        x_rollback = np.hstack((x, temp[:,[0,j]]))
                        x_rollback = np.delete(x_rollback, j, 1)
                        print (regressor_OLS.summary())
                        return x_rollback
                    else:
                        continue
    regressor_OLS.summary()
    return x

SL = 0.05
X_opt = X[:, [0, 1, 2, 3, 4, 5]]
X_Modeled = backwardElimination(X_opt, SL)

